src/models/dirac_mixture_single_publisher_model.py

```diff
@@ -0,0 +1,245 @@
+        """The cross entropy distance between observed and fitted arrays.
+
+        Minimizing cross entropy is equivalent to maximizing likelihood.
+
+        Args:
+            observed_arr: A 1d array of the observed pmf. Or 2d array of the
+                observed pmfs on different directions.
+            fitted_arr: Any other array of pmf(s) that has the same dimension
+                as observed_arr.
+        Returns:
+             The cross entropy of the fitted pmf(s) relative to the observed
+             pmf(s), that is,
+            - sum_i observed_arr[i] * log(fitted_arr[i]) for 1d arrays,
+            - sum_{i, j} observed_arr[i, j] * log(fitted_arr[i, j]) for 2d arrays.
+        """
+        # Small shift to avoid log-zero numerical error.
+        fitted_arr = fitted_arr + 1e-200
+        return -cp.sum(cp.multiply(observed_arr, cp.log(fitted_arr)))
+
+    @staticmethod
+    def solve_optimal_weights(
+        observed_arr: np.ndarray, component_arrs: List, distance: Callable
+    ) -> np.ndarray:
+        """Convex optimization of weights on different components.
+
+        Args:
+            observed_arr: Any 1d or 2d arry.
+            component_arrs: A list of arrays that are of the same dimension
+                of observed_arr.
+            distance: A function on two arrays of the same dimension. Calculates
+                a certain distance between the two arrays.
+
+        Returns:
+            A weight vector such that the weighted sum of the arrays in
+            component_arrs is closest to observed_arr with respect to the given
+            distance.
+        """
+        ws = cp.Variable(len(component_arrs))
+        pred = cp.sum([cp.multiply(w, arr) for w, arr in zip(ws, component_arrs)])
+        problem = cp.Problem(
+            objective=cp.Minimize(distance(observed_arr, pred)),
+            constraints=[ws >= 0, cp.sum(ws) == 1],
+        )
+        # cvxpy is an interface of multiple convex optimization solver packages.
+        # Among them, the open source packages are ECOS
+        # (https://github.com/embotech/ecos/wiki), SCS
+        # (https://www.cvxgrp.org/scs/api/index.html) and OSQP, where ECOS and
+        # SCS are able to handle our problem. So, we try ECOS first and then SCS.
+        # Note: ECOS is available in C.  SCS available in C/C++.  But they are
+        # both unavailable in Java.
+        try:
+            problem.solve()  # The default solver is ECOS
+        except cp.SolverError:
+            problem.solve(solver=cp.SCS)
+        return ws.value
+
+    def fit(self):
+        """Fits a univariate mixed Poisson distribution."""
+        if self.fitted:
+            return
+        self.component_pmfs = [
+            self.truncated_poisson_pmf_vec(component, self.max_freq)
+            for component in self.components
+        ]
+        self.ws = self.solve_optimal_weights(
+            self.observed_pmf,
+            self.component_pmfs,
+            self.cross_entropy,
```

Comment:
It'll be very interesting to compare these two options.  Would you take this AI?  Let me know!

Reply:
@jiayu-google Now that I thought about it, I don't think there is much difference. But I will compare the two during the weekend just in case. Thanks.

Reply:
Thanks, no obligations! Keep me posted and let's collaborate.